x-pack/plugins/files/docs/tutorial.mdx

```diff
@@ -0,0 +1,145 @@
+
+<DocCallOut title="File contents are immutable">
+  Once uploaded, the contents of a file do not change. Browsers, and other clients, should only need to download the contents of a file once, subsequent downloads should be served from the cache.
+
+  If you need to update the contents of a file, you should create a new file with the new contents.
+</DocCallOut>
+
+### How to set up files for your plugin
+
+All setup examples are based on the [`filesExample` plugin](https://github.com/elastic/kibana/blob/d431e87f7ff43833bff085e5bb5b2ab603bfa05d/x-pack/examples/files_example/README.md).
+
+First add the  `files` plugin as a required dependency in your  `kibana.json`:
+
+```json
+  "requiredPlugins": ["files"],
+```
+<br />
+Next define your file kinds. Your plugin can have more than one file kind. Each file kind should represent a specific use case, for example: an image for user avatars.
+
+```ts
+import { FileKind } from '@kbn/files-plugin/common';
+
+export const PLUGIN_ID = 'filesExample';
+
+const httpTags = {
+  tags: [`access:${PLUGIN_ID}`], // ensure that only users with access to this plugin can files of this kind
+};
+
+export const exampleFileKind: FileKind = {
+  id: 'filesExample',
+  maxSizeBytes: 8 * 1024 * 1024, // 8 MiB
+  http: {
+    create: httpTags,
+    delete: httpTags,
+    download: httpTags,
+    getById: httpTags,
+    list: httpTags,
+    // share: httpTags,
+    // update: httpTags,
+  },
+};
+```
+
+In this example we have chosen to disallow metadata updates and file sharing endpoints. **Make sure that you only expose the actions you need over HTTP**.
+
+Now we are able to register this file kind with the file service:
+
+```ts
+  // in your server-side plugin code
+  public setup(core: CoreSetup, { files }: {  files: FilesSetup }) {
+    files.registerFileKind(exampleFileKind);
+  }
+```
+
+You are now able to access your files both from the server-side and the public-side of your plugin. For example, you can now list all of the files of a file kind by doing the following:
+
+```ts
+  // in your public code
+  const result = await files.filesClientFactory.asScoped(exampleFileKind.id).list();
+```
+
+To create a new file from the browser you can do the following:
+
+```ts
+  const { file } = await files.example.create({
+    name: names[Math.floor(Math.random() * names.length)],
+    alt: 'My image',
+    /**
+    * How you define metadata is entirely up to you. Just make sure that your application
+    * remains compatible with previous versions of your metadata.
+    */
+    meta: { myValue: 'test' } as MyMetadata,
+    mimeType: 'image/png',
+  });
+  // ...
+  await files.example.upload({ id: file.id, body: blob });
+```
+
+Now display your file to the world using the client side API:
+
+```tsx
+const { files: [file] } = await fileClient.find({ meta: { myValue: 'test' } });
+
+<img
+    alt={file.alt ?? 'unknown'}
+    src={fileClient.getDownloadHref(file)}
```

Comment:
Maybe it would be useful to construct file URL by just file ID in some scenarios:

```suggestion
    src={fileClient.getFileContentsUrl(fileId)}
```

Reply:
OK, that makes sense to me. I was initially thinking that by requiring the `FileJSON` to be present it's more likely that they fetched the file and that the file exists. But I think your version is more ergonomic: https://github.com/elastic/kibana/issues/139894